server/src/internalClusterTest/java/org/elasticsearch/index/engine/MaxDocsLimitIT.java

```diff
@@ -0,0 +1,212 @@
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.query.MatchAllQueryBuilder;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.IndexShardTestCase;
+import org.elasticsearch.index.translog.Translog;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.plugins.EnginePlugin;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.test.hamcrest.ElasticsearchAssertions;
+import org.junit.After;
+import org.junit.Before;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.Phaser;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.both;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+
+public class MaxDocsLimitIT extends ESIntegTestCase {
+
+    private static final AtomicInteger maxDocs = new AtomicInteger();
+
+    public static class TestEnginePlugin extends Plugin implements EnginePlugin {
+        @Override
+        public Optional<EngineFactory> getEngineFactory(IndexSettings indexSettings) {
+            return Optional.of(config -> {
+                assert maxDocs.get() > 0 : "maxDocs is unset";
+                return EngineTestCase.createEngine(config, maxDocs.get());
+            });
+        }
+    }
+
+    @Override
+    protected boolean addMockInternalEngine() {
+        return false;
+    }
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        List<Class<? extends Plugin>> plugins = new ArrayList<>(super.nodePlugins());
+        plugins.add(TestEnginePlugin.class);
+        return plugins;
+    }
+
+    @Before
+    public void setMaxDocs() {
+        maxDocs.set(randomIntBetween(10, 100)); // Do not set this too low as we can fail to write the cluster state
+        IndexWriterMaxDocsChanger.setMaxDocs(maxDocs.get());
+    }
+
+    @After
+    public void restoreMaxDocs() {
+        IndexWriterMaxDocsChanger.restoreMaxDocs();
+    }
+
+    public static void assertNoOutstandingReservingDocs() throws Exception {
+        assertBusy(() -> {
+            for (IndicesService indicesService : internalCluster().getDataNodeInstances(IndicesService.class)) {
+                for (IndexService indexService : indicesService) {
+                    for (IndexShard indexShard : indexService) {
+                        try {
+                            assertThat(EngineTestCase.getReservingDocs(IndexShardTestCase.getEngine(indexShard)), equalTo(0L));
```

Comment:
Can we make this assertion more generally too? IIRC there are some other assertions that check that things have settled down after a test (no in-flight operations, for instance) although I'm not quite sure where they are.

Reply:
This assertion should be very lightweight. I added it in https://github.com/elastic/elasticsearch/pull/63273/commits/290806d102a5f63ecd2af1476ba0fffb360ffc77.